File: csharp/AdventOfCode.2021.Console/Day01.cs
namespace AdventOfCode._2021.Console;

internal class Day01 : IDay
{
    public int DayNumber => 1;
    public int PartOneTestAnswer => 7;
    public int PartTwoTestAnswer => 5;

    public int ExecutePartOne(string[] input)
    {
        var inputAsIntegers = input.Select(int.Parse).ToArray();
        return WindowList(inputAsIntegers, 2).Count(x => x.First() < x.Last());
    }

    public int ExecutePartTwo(string[] input)
    {
        var inputAsIntegers = input.Select(int.Parse).ToArray();
        return WindowOfWindowList(WindowList(inputAsIntegers, 3)).Count(x => x.First().Sum() < x.Last().Sum());
    }

    static IEnumerable<int[]> WindowList(IReadOnlyList<int> list, int windowSize)
    {
        var newList = new List<int[]>();
        for (var i = windowSize - 1; i < list.Count; i++)
        {
            var listOfValues = new List<int>();
            for (var j = windowSize - 1; j >= 0; j--)
            {
                listOfValues.Add(list[i - j]);
            }

            newList.Add(listOfValues.ToArray());
        }

        return newList;
    }

    static IEnumerable<int[][]> WindowOfWindowList(IEnumerable<int[]> list)
    {
        var newList = new List<int[][]>();
        var intsEnumerable = list as int[][] ?? list.ToArray();
        for (var i = 1; i < intsEnumerable.Length; i++)
        {
            newList.Add(new[] { intsEnumerable[i - 1], intsEnumerable[i] });
        }

        return newList;
    }
}

File: typescript/2021/day01.ts
import { Day } from "../program";

export class Day01 implements Day {
  partOneTestAnswer = 7;
  partTwoTestAnswer = 5;

  executePartOne(input: string[]): number {
    const numbers = input.map((x) => +x);

    return Day01.getCountOfIncreasesInArray(numbers);
  }

  executePartTwo(input: string[]): number {
    const numbers = input.map((x) => +x);
    const slidingWindow = [];
    for (let i = 2; i < numbers.length; i++) {
      slidingWindow.push(numbers[i - 2] + numbers[i - 1] + numbers[i]);
    }

    return Day01.getCountOfIncreasesInArray(slidingWindow);
  }

  private static getCountOfIncreasesInArray(numbers: number[]) {
    let increaseCount = 0;
    for (let i = 1; i < numbers.length; i++) {
      if (numbers[i] > numbers[i - 1]) {
        increaseCount++;
      }
    }

    return increaseCount;
  }
}
